criteria.c

```diff
@@ -456,3 +448,34 @@ struct mako_criteria *create_criteria_from_notification(
 
 	return criteria;
 }
+
+
+// To keep the behavior of criteria predictable, there are a few rules that we
+// have to impose on what can be modified depending on what was matched:
+// - Criteria matching grouped notifications are not allowed to change the
+//   anchor, output, or group-by, as this would invalidate the grouping.
+bool validate_criteria(struct mako_criteria *criteria) {
+	if (criteria->spec.grouped || criteria->spec.group_index) {
+		static const char *message =
+			"Setting `%s` is not allowed when matching `grouped` or "
+			"`group-index`\n";
+
+		if (criteria->style.spec.anchor) {
+			fprintf(stderr, message, "anchor");
+			return false;
+		} else if (criteria->style.spec.output) {
+			fprintf(stderr, message, "output");
+			return false;
+		} else if (criteria->style.spec.group_criteria_spec) {
+			fprintf(stderr, message, "group-by");
+			return false;
+		}
```

Comment:
Maybe this would be slightly better?

```c
const char *invalid_field = NULL;
if (criteria->style.spec.anchor) {
    invalid_field = "anchor";
} else if (…) {
    …
}
if (invalid_field) {
    fprintf(stderr, "Setting …", invalid_field);
    return false;
}
```

Compilers don't usually like it when the format string is stored in a variable.

Reply:
Yep, I like that a lot better.